### DISCUSSION TITLE
Expected format for SMTP send responses

### DISCUSSION THREAD

**Discussion by lonix1** [jstedfast/MailKit#1842] [Q&A] [answered]
The `SendAsync` returns a string described as *"The final free-form text response from the server"*. I assume it varies between servers.

I've observed:
- When using a service like AWS SES, one receives JSON containing a "message id" (a UUID) which can be later correlated to delivery / bounce / complaint notifications.
- Other providers use different approaches, for example, a string with "OK id=SOME_ID_OR_UUID".

I believe SMTP servers respond with various status codes - are those returned somehow? Is there a standard governing that final response (for example, the "OK" or the code/uuid)? If not, is there a general format one expects to find for "larger" mail providers such as AWS, GMail, Outlook, SendGrid, MailGun, Postmark, etc.? Or perhaps some samples of such responses?

(I'm trying to make our code more resilient, and to be able to correlate mails to ultimate notifications, but the response's unstructured format is obviously tricky.)

Any pointers would be appreciated!

**Comment by jstedfast** [comment, marked as answer]
Unfortunately, there is no standard which is why I had to return a string of free-form text.

**Reply by lonix1** [reply]
Thanks, I thought so, but hoped I was wrong.

In your experience, do mail providers use radically different texts or is there a sane general format they use? From what I've seen so far - but I haven't seen as many as you - that response isn't predictable at all.

I've noticed [various SMTP codes](https://en.wikipedia.org/wiki/List_of_SMTP_server_return_codes), and even those are sometimes returned and sometimes not.

**Reply by jstedfast** [reply]
Your observations match mine exactly.

AWS, Azure, GMail and maybe others will typically have some sort of tracking id that is likely to be tied into some sort of service telemetry for these big organizations.

ISP mail servers may or may not have this info depending on what software they use. Bigger ISP's like Comcast are probably more likely to have some sort of tracking value, but I'm not sure.
